### PR TITLE
[SPARK-5754] [yarn] Spark/Yarn/Windows driver/executor escaping Fix

### PR DESCRIPTION
**Pull request by cbvoxel**
This is my retry to suggest a fix for using Spark on Yarn on Windows. The former request lacked coding style which I hope to have learned to do better, and wasn't a true solution as I didn't really understand where the problem came from. Albeit being still a bit obscure, I can name the "players" and have come up with a better explaination of why I am suggesting this fix. 

I also used @vanzin and @srowen input to _try_ to give a more elegant solution. I am not so sure if that worked out though.

I still hope that this PR is a lot more useful than the last. Also do I hope that this is a _solution_ to the problem that Spark doesn't work on Yarn on Windows. With these changes it works (and I can also explain why!).

I still believe that a Unit Test should be included, kind of like the one I committed the last time. But that was premature, as I want to get the principal 'Go' from @vanzin and @srowen. 

Thanks for your time both of you.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 1]
```diff
@@ -21,6 +21,8 @@ import java.io.File
 import java.util.regex.Matcher
 import java.util.regex.Pattern
 
+import org.apache.spark.launcher.YarnCommandBuilderUtils
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: nit: this import should be grouped with other `org.apache.spark` imports.

File: yarn/src/main/scala/org/apache/spark/launcher/YarnCommandBuilderUtils.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Exposes needed methods
+ * TODO: Is there a better way?
+ */
+private[spark] object YarnCommandBuilderUtils {
+  def quoteForBatchScript(arg: String) : String = {
+    org.apache.spark.launcher.CommandBuilderUtils.quoteForBatchScript(arg)
+  }
+
+  def quoteForCommandString(arg: String) : String = {
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/launcher/YarnCommandBuilderUtils.scala]: This is not used anywhere.

File: yarn/src/main/scala/org/apache/spark/launcher/YarnCommandBuilderUtils.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    org.apache.spark.launcher.CommandBuilderUtils.quoteForCommandString(arg)
+  }
+
+  def isWindows : Boolean = {
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/launcher/YarnCommandBuilderUtils.scala]: You don't really need this. There's already a `org.apache.spark.util.Utils.isWindows` that you can use.

File: yarn/src/main/scala/org/apache/spark/launcher/YarnCommandBuilderUtils.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+package org.apache.spark.launcher
+
+/**
+ * Exposes needed methods
+ * TODO: Is there a better way?
+ */
+private[spark] object YarnCommandBuilderUtils {
+  def quoteForBatchScript(arg: String) : String = {
+    org.apache.spark.launcher.CommandBuilderUtils.quoteForBatchScript(arg)
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/launcher/YarnCommandBuilderUtils.scala]: `CommandBuilderUtils` is in the same package, so you don't need the fully-qualified name here.

File: yarn/src/main/scala/org/apache/spark/launcher/YarnCommandBuilderUtils.scala [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.launcher
+
+/**
+ * Exposes needed methods
+ * TODO: Is there a better way?
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/launcher/YarnCommandBuilderUtils.scala]: Not really without exposing these methods in the launcher library (which we don't want to). BTW you can remove the comment.

**Comment by vanzin** [comment]
This looks like it's going in the right direction. In your previous patch, you mentioned that `quoteForBatchScript` didn't escape `%` characters; is that not needed?

Also, could you write a better PR title? It should explain what the change is doing, not what the problem was.

**Comment by cbvoxel** [comment]
@vanzin: no it doesn't seem to be needed. It seems that `quoteForBatchScript` works correctly with arguments entered on the command line. Problems arise with called cmd files though.
What I see is that in cmd files every `%` needs to be escaped by another `%`. This really doesn't happen on the command line! I have an example (use of grep: I have MinGW msys installed):

### Using the command line ###
> java "-XX:OnOutOfMemoryError=taskkill /F /PID %%%%p" -XX:+PrintFlagsFinal -version | grep OnOutOfMemoryError
>     bool HeapDumpOnOutOfMemoryError                = false           {manageable}
> ccstrlist OnOutOfMemoryError                       := taskkill /F /PID %%%%p{product}
> java version "1.7.0_79"
> Java(TM) SE Runtime Environment (build 1.7.0_79-b15)
> Java HotSpot(TM) 64-Bit Server VM (build 24.79-b02, mixed mode)

### Using the a .cmd file ###
> sparkCmdTest.cmd
>     bool HeapDumpOnOutOfMemoryError                = false           {manageable}
> ccstrlist OnOutOfMemoryError                       := taskkill /F /PID %%p{product}
> java version "1.7.0_79"
> Java(TM) SE Runtime Environment (build 1.7.0_79-b15)
> Java HotSpot(TM) 64-Bit Server VM (build 24.79-b02, mixed mode)

where `sparkCmdTest.cmd` has this one line:

> java "-XX:OnOutOfMemoryError=taskkill /F /PID %%%%p" -XX:+PrintFlagsFinal -version | grep OnOutOfMemoryError

Now what made me wonder is, that for the executor to work we somehow need four (!) `%`. I would 'guess' that somewhere along the line there is another cmd file being called, and that with the first pass `%%%%` is reduced to `%%` and then to `%`. Acutally I have no solid idea, all I can say that if I executed the cmd file spark passes to Yarn, it *must* be four `%`.

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
I can't really test this, but since the Linux tests are passing, LGTM. I'll merge this Monday morning unless someone else says something.

**Comment by vanzin** [comment]
One last time jenkins, retest this please.

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
Alright, YarnClusterSuite should be fixed now, so let's try this again. retest this please

**Comment by vanzin** [comment]
Ok, merging this.

**Comment by vanzin** [comment]
Merged to master, thanks!

**Pull request closed without merging**